NE-542 Expose router compression variables (#887)
openshift/enhancements: enhancements/ingress/haproxy-router-compression.md

their ingress workloads, as described in feature request [NE-542](https://issues.redhat.com/browse/NE-542).
HAProxy in OpenShift 3.x exposed compression variables, so lack of access to the variables in OpenShift 4.x
is a migration blocker for customers that were using them in OpenShift 3.x.

The ability to configure compression was available in 3.11 and disappeared in 4.x. This proposal
reintroduces this ability to the extent of feature parity with 3.11.

### Goals

Enable the configuration of HTTP compression global defaults via the `IngressControllerSpec`, which will
expose the two environment variables `ROUTER_ENABLE_COMPRESSION` and `ROUTER_COMPRESSION_MIME`.

Enable HTTP compression on selected MIME types, be they wildcard (e.g. `text/*`, `application/*`, etc.) or
specific (e.g. `text/html`, `text/plain`, `text/css`, `application/json`, `image/svg+xml`, `font/eot`, etc.).

By exposing the environment variables, any HTTP request that terminates within HAProxy, with the header
**"Accept-Encoding: gzip"**, and with a MIME type that matches one of the configured MIME types will qualify to have

> HTTP compression applied to its payload.  Note 1: this applies only to edge-terminated and reencrypt routes.  Passthrough

Author: Yes, updated it

//
// This field is optional, and its absence implies that compression should not be enabled
// globally in HAProxy.
//
// If httpCompressionPolicy exists, compression should be enabled only for the specified
// MIME types
type HTTPCompressionPolicy struct {
	// compressionMIMETypes is a list of MIME types that should have compression applied.
	// At least one MIME type must be specified.
	//
	// Note: Not all MIME types benefit from compression, but HAProxy will still use resources
	// to try to compress if instructed to.  Generally speaking, text (html, css, js, etc.)
	// formats benefit from compression, but formats that are already compressed (pdf, image,
	// audio, video, etc.) benefit little in exchange for the time and cpu spent on compressing
	// again. See https://joehonton.medium.com/the-gzip-penalty-d31bd697f1a2
	//
	// +required
	// +kubebuilder:validation:MinItems=1
	CompressionMIMETypes []CompressionMIMEType `json:"compressionMIMETypes,omitempty"`
}

// CompressionMIMEType defines the format of a single MIME type
// E.g. "text/css; charset=utf-8", "text/html", "text/*", "image/svg+xml",
// "application/octet-stream", etc.
//

> // +kubebuilder:validation:Pattern="[a-zA-Z0-9\\+-;= ]+/([a-zA-Z0-9\\\+-;= ]+|\\*)"

Reviewer: This should be anchored.  You shouldn't need to escape `+` in a character class.  Are spaces really allowed in the type?  https://datatracker.ietf.org/doc/html/rfc1341#page-7 indicates that spaces, `;`, and `=` should be prohibited in the type and subtype.

Author: "text/css; charset=utf-8" is a valid MIME type, so I included spaces, `;`, and `=` in the pattern.

Author: And will add `"`.

Author: I kept this pretty loose.  RFC1341 is made obsolete by https://datatracker.ietf.org/doc/html/rfc1521#page-68  

I will anchor and remove escapes from `+`, but if we don't keep this loose, we need to use https://datatracker.ietf.org/doc/html/rfc1521#page-68

type CompressionMIMEType string
```

To set up the IngressControllerSpec to use HTTP compression for a set of specific MIME types, the configuration
might resemble:
```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpCompression:
    compressionMIMETypes:
    - "text/html"
    - "text/css; charset=utf-8"
    - "application/json"
```

> To expose the HTTP compression variables, further changes will be added to the ingressi router deployment:

Reviewer: Looks like this got cut off.  
```suggestion
To expose the HTTP compression variables, further changes will be added to the ingress operator's logic to build the router deployment:
```

```go
const (
	WildcardRouteAdmissionPolicy = "ROUTER_ALLOW_WILDCARD_ROUTES"
    ...
	RouterEnableCompression = "ROUTER_ENABLE_COMPRESSION"
	RouterCompressionMIMETypes = "ROUTER_COMPRESSION_MIME"
)
...
    // desiredRouterDeployment returns the desired router deployment.
    func desiredRouterDeployment(ci *operatorv1.IngressController, ingressControllerImage string, ingressConfig *configv1.Ingress, apiConfig *configv1.APIServer, networkConfig *configv1.Network, proxyNeeded bool, haveClientCAConfigmap bool, clientCAConfigmap *corev1.ConfigMap) (*appsv1.Deployment, error) {
    ...
      if len(ci.Spec.HTTPCompression.CompressionMIMETypes) != 0 {
        env = append(env, corev1.EnvVar{Name: RouterEnableCompression, Value: "true"})

>         env = append(env, corev1.EnvVar{Name: RouterCompressionMIMETypes, Value: string.Join(spec.httpCompression.compressionMimeTypes, " ")})

Reviewer: Do we need to quote or escape values with spaces (such as `text/css; charset=utf-8` in the example above)?

Author: The quotes need to be there already.  No validation done at this point.  I'll add quotes to the pattern.

Author: They need to be quoted, but I would leave that up to the user.  Wdyt?

compression directives will become no-op: HAProxy will see the compressed response and will not compress again. If a
backend server does not support HTTP compression, and the header "Accept-Encoding" exists in the request, HAProxy will
compress the matching response.  Also, if the HAProxy doesn't have enough resources to perform compression, it may skip
the compression in order to stay functional.  Process-level directives exist to tune this behavior: `maxcomprate`,
`maxcompcpuusage`, and `tune.comp.maxlevel` as described in the
[HAProxy compression documentation](https://www.haproxy.com/documentation/hapee/latest/load-balancing/compression/),
but these are not in scope of this proposal.

### Risks and Mitigations

Overuse of compression can cause the HAProxy to devote more resources to compression than to other operations,
compromising its operation.  We do not have any evidence that this will occur, and give detailed notes on use of
necessary MIME types to mitigate pointless compression and overuse of compression.  Should this become a problem
in the future, we can add default values for the process-level directives mentioned above (`maxcomprate`,

> `maxcompcpuusage`, and `tune.comp.maxlevel`)  The team consensus is that adding more tunable knobs at this stage is

Reviewer: ```suggestion
`maxcompcpuusage`, and `tune.comp.maxlevel`).  The team consensus is that adding more tunable knobs at this stage is
```

## Design Details
Design details are covered in the Proposal section, but some additional detailed notes from the
[HAProxy compression documentation](https://www.haproxy.com/documentation/hapee/latest/load-balancing/compression/)
are listed below.

### Compression types
MIME types to apply compression to,  are read from the Content-Type response header field. If a MIME type is not set,
HAProxy tries to compress all responses that are not compressed by the server.  Therefore, at least one MIME type is
required in the `HTTPCompressionPolicy`.  We consider the fact that cluster admins may want to create and use their
own MIME type, but the allowed set of characters is defined in a kubebuilder validation
pattern  of `"[a-zA-Z0-9\\+-;= ]+/([a-zA-Z0-9\\\+-;= ]+|\\*)"`, to allow a token of alphanumeric plus "_-;= " followed
by "/" followed by another alphanumeric token of the same pattern, or "*", the wildcard character.

### Open Questions
Is the MIME type pattern still too strict?  I reviewed the `mime` golang package at https://golang.org/src/mime/type.go,
`builtinTypesLower` map, but if we intend to allow any custom MIME type, perhaps a wider net should be cast, as in

> https://golang.org/src/mime/grammar.go.

Reviewer: https://datatracker.ietf.org/doc/html/rfc1341#page-7 provides BNF definitions for allowed MIME types.  If we want to make it a little looser (e.g., allowing non-standard types without the `x-` prefix), I don't object.

Author: https://datatracker.ietf.org/doc/html/rfc1521#page-68 is more current.  Should we go that far?

  the generated haproxy.cfg
  - that only requested MIME types receive compression

Note that HAProxy Enterprise **disables** HTTP compression when:
- The request does not advertise a supported compression algorithm in the "Accept-Encoding" header
- The response message is not HTTP/1.1 or above
- The HTTP status code is not one of 200, 201, 202 or 203
- The response contains neither a "Content-Length" header, nor a "Transfer-Encoding" header whose last value is `chunked`
- The response contains a "Content-Type" header whose first value starts with `multipart`
- The response contains a "Cache-Control" header with a value of `no-transform`
- The response contains a "Content-Encoding" header, indicating that the response is already compressed (see `compression offload`)
- The response contains an invalid "ETag" header or multiple "Etag" headers

### Graduation Criteria

This feature will be immediately available in release 4.10, and will not have a dev or tech preview.

#### Dev Preview -> Tech Preview
n/a

#### Tech Preview -> GA
n/a

#### Removing a deprecated feature

- Announce deprecation and support policy of the existing feature

> - Deprecate the feature

Reviewer: N/A.

### Upgrade / Downgrade Strategy

An upgrade from a previous installation that had the compression directives exposed, will need to change to use the

> IngressController spec.

Reviewer: Previous versions did not expose the compression directives.  The API is additive and turned off by default, so there should be no change in behavior or need for migration during upgrades.

Author: Fixed.

environment variables is not exposed.
```go
defaults
...
{{- if isTrue (env "ROUTER_ENABLE_COMPRESSION") }}
  compression algo gzip
  compression type {{ env "ROUTER_COMPRESSION_MIME" "text/html text/plain text/css" }}
  {{- end }}
```

### Global HTTP Compression via IngressControllerSpec and environment variables

The proposed change to IngressControllerSpec exposes the HTTP compression configuration variables at the default level:
```go
type IngressControllerSpec struct {
...
    // httpCompression defines policy for HTTP traffic compression
    // 
    // The default value is no compression
    //
    // +optional
    HTTPCompression HTTPCompressionPolicy `json:"httpCompression,omitempty"`
}
...
// httpCompressionPolicy turns on compression for the specified MIME types
//
// This field is optional, and its absence implies that compression should not be enabled
// globally in HAProxy.
//
// If httpCompressionPolicy exists, compression should be enabled only for the specified
// MIME types
type HTTPCompressionPolicy struct {
	// compressionMIMETypes is a list of MIME types that should have compression applied.

> 	// At least one MIME type must be specified.

Reviewer: ```suggestion
	// This list can be empty, in which case the ingress controller does not apply compression.
```

	//
	// Note: Not all MIME types benefit from compression, but HAProxy will still use resources
	// to try to compress if instructed to.  Generally speaking, text (html, css, js, etc.)
	// formats benefit from compression, but formats that are already compressed (pdf, image,
	// audio, video, etc.) benefit little in exchange for the time and cpu spent on compressing
	// again. See https://joehonton.medium.com/the-gzip-penalty-d31bd697f1a2
	//
	// +required
	// +kubebuilder:validation:MinItems=0

> 	CompressionMIMETypes []CompressionMIMEType `json:"compressionMIMETypes,omitempty"`

Reviewer: Maybe we could delete "compression" in the field name to avoid stutter (`spec.httpCompression.compressionMIMETypes` → `spec.httpCompression.mimeTypes`):
```suggestion
	CompressionMIMETypes []CompressionMIMEType `json:"mimeTypes,omitempty"`
```

}

// CompressionMIMEType defines the format of a single MIME type
// E.g. "text/css;charset=utf-8", "text/html", "text/*", "image/svg+xml",
// "application/octet-stream", "X-custom/customsub", etc.
// The format should follow the Content-Type definition in RFC1341 - https://datatracker.ietf.org/doc/html/rfc1341#page-7
//

> // +kubebuilder:validation:Pattern="^"?(X\-[^ \(\)<>@,;:\\"\/\[\]\?.\=]+|[application|audio|image|message|multipart|text|video])"?\/"?([^ \(\)<>@,;:\\"\/\[\]\?.\=]+(;? *([^ \(\)<>@,;:\\"\/\[\]\?.\=]+=([^ \(\)<>@,;:\\"\/\[\]\?.\=]|".+")+))?)+"?$

Reviewer: The quoting with double-quotes looks strange, you don't need to escape things inside brackets, you do need to put any closing bracket first inside brackets (e.g., `[^][ ()<>@,;:"/?.=]`), and you probably want parentheses instead of brackets here: `[application|audio|image|message|multipart|text|video]`.  Do we need to allow double-quotes in the provided MIME type?  Overall, trying to do this validation with a regexp might be more effort than it's worth; it might be better to make things looser, even if that would allow some things that the RFC doesn't strictly allow (such as arbitrary types), but prohibit redundant quoting (if I understand correctly, the yaml syntax might have double-quotes, but the value that is expressed in yaml shouldn't, and the operator should add double-quotes in the generated HAProxy configuration if they are needed there).

Author: I'm going to go with this pattern: `^"?(X-[^ ()[\]<>@,;:"/?.=]+|(application|audio|image|message|multipart|text|video))"?/"?(.+)+"?$`

required in the `HTTPCompressionPolicy`.  We consider the fact that cluster admins may want to create and use their
own MIME type, but the allowed set of characters is defined in a kubebuilder validation
pattern  of `"[a-zA-Z0-9\\+-;= ]+/([a-zA-Z0-9\\\+-;= ]+|\\*)"`, to allow a token of alphanumeric plus "_-;= " followed
by "/" followed by another alphanumeric token of the same pattern, or "*", the wildcard character.

### Open Questions
n/a

### Test Plan
- Unit tests will validate
  - that at least one HTTP compression MIME type exists and any listed MIME types adhere to the
  correct pattern
  - that the HTTP compression directive appears in the `defaults` section of the HAProxy template
  when it is enabled and properly configured on the IngressController spec
  - that a downgrade to version 4.9 (or version previous to feature implementation version) will not retain the
  compression configuration of 4.10 (the environment variables should no longer be exposed)
- E2e tests will validate
  - that the IngressController spec accepts and enables compression in the `defaults` section of
  the generated haproxy.cfg
  - that only requested MIME types receive compression

> Note that HAProxy Enterprise **disables** HTTP compression when:

Reviewer: We use HAProxy (not HAProxy Enterprise), but I think both have the same behavior.

    haproxy.router.openshift.io/compressionMimeTypes: "text/html text/css application/pdf"
````

### User Stories

#### As a cluster administrator, I need to enable gzip HTTP compression globally for a selection of MIME types

Select the necessary MIME types and edit the IngressController spec as shown:

```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpCompression:
    compressionAlgorithm: gzip
    compressionMIMETypes:
    - "text/html"
    - ...
```
#### As a cluster administrator, I need to enable gzip HTTP compression for only one route and one MIME type, "text/html"

To apply HTTP compression to all routes in the cluster, use a command like this. (This command requires
an updated `oc` client that supports the `--all-namespaces` flag on the `oc annotate` verb):
```shell
$ oc annotate route --all --all-namespaces --overwrite=true "haproxy.router.openshift.io/compressionAlgorithm"="gzip"
$ oc annotate route --all --all-namespaces --overwrite=true "haproxy.router.openshift.io/compressionMIMETypes"="text/html text/css text/xml"

```

> To apply HSTS to all routes in a particular namespace, use a command like this:

Reviewer 2: s/HSTS/HTTP

Author: done

is a migration blocker for customers that were using them in OpenShift 3.x.

The ability to configure compression was available in 3.x, but it applied only to the HAProxy
`defaults` configuration section, and supported only the `gzip` compression algorithm.  This proposal
reintroduces and then enhances the previous implementation in these ways:
- allow the `frontend`, and `backend` sections in addition to the `defaults` section, which would provide for more
  fine-grain configuration at the route level.
- allow a choice of `gzip` or `raw-deflate` compression algorithms.

### Goals

Enable the configuration of HTTP compression global defaults via the `IngressControllerSpec`.

Enable the configuration of HTTP compression for route backends/frontends via route annotations.

Enable HTTP compression on selected MIME types, be they wildcard (e.g. text/*, application/*, etc.) or
specific (e.g. text/html, text/plain, text/css, application/json, image/svg+xml, font/eot, etc.).  Not all MIME
types benefit from compression, but there is a short list found [here](https://letstalkaboutwebperf.com/en/gzip-brotli-server-config/)
that will be implemented as a part of this proposal.

> By exposing ROUTER_ENABLE_COMPRESSION, any HTTP request with the header "Accept-Encoding: gzip" or

Reviewer 3: This will only apply to routes that the router terminates right?  So not pass through.

Author: Thanks, I clarified that.

        return supplied
    }
	return ""
}
```

The proposed change to IngressControllerSpec exposes the HTTP compression configuration variables at the default level:
```go
type IngressControllerSpec struct {
...
    // httpCompression defines policy for HTTP traffic compression
    // 
    // The default value is no compression
    //
    // +optional
    HTTPCompression *HTTPCompressionPolicy `json:"httpCompression,omitempty"`
}
...
// httpCompressionPolicy defines the compression algorithm and MIME types that need to
// have compression applied.
//
// This field is optional, and its absence implies that compression should not be enabled
// globally in HAProxy.
//
// If httpCompressionPolicy exists, compression should be enabled only for the listed
// MIME types, and only using the designated compression algorithm.
type HTTPCompressionPolicy struct {

	// compressionAlgorithm defines the desired compression algorithm.
	//
	// +required
	CompressionAlgorithm CompressionAlgorithmType `json:"compressionAlgorithm,omitempty"`
	
	// compressionMIMETypes is a list of MIME types that should have compression applied
	//
	// +kubebuilder:MinItems=1
	// +required

> 	CompressionMIMETypes []string `json:"compressionMIMETypes,omitempty"`

Reviewer 3: My lack of http knowledge shows here, but is this possible to make an enum to avoid accidents?

Author: Based on feedback from the team, we will not be adding any compressionMIMEType validation or enforcing policy, except maybe a basic regex.

Author: Now I have an open question about whether my latest solution is too strict.

}

// CompressionAlgorithmType defines the type of compression algorithm
// +kubebuilder:validation:Enum=gzip;raw-deflate
type CompressionAlgorithmType string
const (
	gzipCompressionAlgorithm CompressionAlgorithmType = "gzip"
	rawdeflateCompressionAlgorithm CompressionAlgorithmType = "raw-deflate"
)
```

To set up the IngressControllerSpec to use HTTP compression on all routes, the configuration
might resemble:
```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpCompression:
    compressionAlgorithm: gzip
    compressionMIMETypes:
    - "text/html"
    - "text/css"
    - "application/pdf"
```

> ### Route-based HTTP Compression via Annotations

Reviewer 3: Can you describe in this section whether the per-route configuration always overrides the per-ingress configuration?  How does a route indicate no compression when the ingress configuration says to compress?

Author: Based on feedback from the team, we will not add route-based HTTP compression options in this release.

> To expose these variables at the route level, the mechanism will be two route annotations that are processed in the

Reviewer 3: The annotation based API for HSTS is a historical mistake.  Knowing what we know now, it is a better idea to modify the route API itself to add a new stanza.  Probably even the same stanza you have for the ingress spec.

Author: Based on feedback from the team, we will remove route-based HTTP compression options in this release.

- The response contains neither a "Content-Length" header, nor a "Transfer-Encoding" header whose last value is `chunked`
- The response contains a "Content-Type" header whose first value starts with `multipart`
- The response contains a "Cache-Control" header with a value of `no-transform`
- The response contains a "Content-Encoding" header, indicating that the response is already compressed (see `compression offload`)
- The response contains an invalid "ETag" header or multiple "Etag" headers

### Graduation Criteria

This feature will be immediately available in release 4.10, and will not have a dev or tech preview.

#### Dev Preview -> Tech Preview
n/a

#### Tech Preview -> GA
n/a

#### Removing a deprecated feature

- Announce deprecation and support policy of the existing feature
- Deprecate the feature

### Upgrade / Downgrade Strategy

An upgrade from a previous installation that had the compression directives exposed, will need to change to use the
IngressController spec.  No previous installation would have any compression directives via route annotations.

If a cluster administrator applied any of the compression configuration options

> in 4.10, then downgraded to 4.9, the compression configuration settings would no longer be in effect unless they

Reviewer 3: Please be sure that there is an explicit test of this capability done by QE on rollback.

Author: Will add a unit test too.

used a custom unmanaged HAProxy configuration.

### Version Skew Strategy
n/a

## Implementation History
n/a

## Drawbacks

By implementing this proposal we are giving the user tools that could threaten the operation of HAProxy.
Improper configuration of the MIME types or general over-use of compression could allow the HAProxy to
dedicate more processing time to compression than to its main duties of routing and traffic management.

> ## Alternatives

Reviewer 3: Because this feature is not security related and because the side-effects are minimal, short-lived, and fixable by the server enforcement by ingress is ok.  

Does HAProxy allow configuration of a default that applies by domain?  If so, is that more natural for customers to configure?  An answer down here in alternatives would help future us remember why the choice was made.

Author: Based on feedback from the team, we will not add route-based HTTP compression options in this release.  HAProxy does not allow configuration that applies by domain.

    HTTPCompression *HTTPCompressionPolicy `json:"httpCompression,omitempty"`
}
...
// httpCompressionPolicy turns on compression for the specified MIME types
//
// This field is optional, and its absence implies that compression should not be enabled
// globally in HAProxy.
//
// If httpCompressionPolicy exists, compression should be enabled only for the specified
// MIME types
type HTTPCompressionPolicy struct {

	// compressionEnabled defines the desired compression enablement
	// If true, compression is enabled for all MIME types specified in
	// CompressionMIMETypes.  If false, compression is disabled for any MIME type.
	//
	// +required
	// +kubebuilder:default=false

> 	CompressionEnabled bool `json:"compressionEnabled"`

Reviewer 2: Do we need this if we have entries in `CompressionMIMETypes`. If `CompressionMIMETypes` is empty then compression is disabled. I don't know whether users would want the convenience of toggling `CompressionEnabled` vis-a-vis just removing the list of mime types. Thoughts?

Author: I agree, we don't need this. Removed.  @frobware @miciah PTAL at latest change (minor) and tag if lgty.
